### PR TITLE
Unstructured: Implement no-mesh concept and improvements from hackathon

### PR DESCRIPTION
**Pull request by havogt**
In the "no-mesh" description, the mesh argument is replaced by an explicit reference to a connectivity, e.g.

old syntax
```
dtype = DataType.FLOAT64

def sten(mesh: Mesh, field_in: Field[Vertex, dtype], field_out: Field[Edge, dtype]):
    with computation(FORWARD), location(Edge) as e:
        field_out[e] = sum(field_in[v] for v in vertices(e))
```

new syntax
```
E2V = types.new_class("E2V", (Connectivity[Edge, Vertex, 2, False],))
dtype = DataType.FLOAT64

def sten(e2v: E2V, field_in: Field[Vertex, dtype], field_out: Field[Edge, dtype]):
    with computation(FORWARD), location(Edge) as e:
        field_out[e] = sum(field_in[v] for v in e2v[e])
```

Changes to frontend:
- Python AST Node Capture now supports invertable transformations of Capture nodes, e.g. a captured ast.argument.name can be transformed into an ast.name
- Added support for native functions (max, min, mod, sqrt)
- Enhanced subclass check for `BuiltInType`s
- Added support for Connectivities, SparseField (including assignments and temporaries), LocalFields (used for weighted reductions)
- Frontend now uses the eve supplied symbol table (types are now declared using declerations contained in the AST)
- References to BuiltInTypes used inside stencils (dtype, Vertex, ..., Connectivity, LocalField) are now supplied using Externals
- Added basic support for Fields with a vertical dimension (hacky & unfinished)
- Fixed pydantic type coercion problems in the frontend
- Bugfix for negative literals / constants
- Added poor-mans const expression evaluator (currently only evaluates constants and type expressions, hacky)
- Added preliminary type deduction mechanism (deducing the type of an expression, hacky)
- Added tests for k-dimension-only-fields, native functions, sparse field assignments, weights / LocalFields

Changes to IRs + code generators
- Introduce the Connectivity concept
- NeighborChains are replaced by references to connectivities (or domain)
- Extended IRs for the new concepts developed during hackathon: assignments to SparseFields, proper treatment of field dimensionality.
- Updated the code generator to work without the mesh concept and using simple helpers

Changes to C++ helpers
- Replaced "gridtools::next" mesh functions by simple helpers "gridtools::usid""

### CODE REVIEW

File: src/gtc_unstructured/frontend/py_to_gtscript.py [hunk 1]
```diff
@@ -202,26 +260,45 @@ def transform(self, node, eligible_node_types=None):
...
                                 capture, eligible_capture_types
                             )
-                    return node_type(**transformed_captures)
+
+                    assert len(args) + len(transformed_kwargs) == len(captures)
+
+                    try:
+                        node_type(*args, **transformed_kwargs)
+                    except:
+                        bla = 1 + 1
```
Comment (tehrengruber) [inline src/gtc_unstructured/frontend/py_to_gtscript.py:281]:
```suggestion
```

File: tests/gtc_unstructured_tests/regression/cpp/fortran/vertex2edge_wrapper.cu [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+template <class T>
+void run_stencil_impl(stencil_t stencil, fortran_array_view<T, 1> in,
+                      fortran_array_view<T, 1> out) {
+  stencil(in, out);
+}
+
+BINDGEN_EXPORT_BINDING_WRAPPED(2, alloc_stencil, alloc_stencil_impl);
+BINDGEN_EXPORT_GENERIC_BINDING_WRAPPED(3, run_stencil, run_stencil_impl,
+                                       (double)(float));
+} // namespace icon
```
Comment (tehrengruber) [inline tests/gtc_unstructured_tests/regression/cpp/fortran/vertex2edge_wrapper.cu:33]: @havogt Should we use a symlink to the .cc file here?

**Review by tehrengruber — APPROVED** [review]
TODO

**Comment by havogt** [comment]
bors try

**Pull request merged**